src/tabmat/formula.py

```diff
@@ -0,0 +1,313 @@
+            return self.multiply(other)
+        elif isinstance(
+            other, (InteractableSparseMatrix, InteractableCategoricalMatrix)
+        ):
+            return other.__mul__(self)
+        else:
+            raise TypeError(f"Cannot multiply {type(self)} and {type(other)}")
+        # Multiplication with sparse and categorical is handled by the other classes
+
+    def __rmul__(self, other):
+        return self.__mul__(other)
+
+    def to_non_interactable(self):
+        return DenseMatrix(self)
+
+    def get_names(self, col):
+        return [col]
+
+
+class InteractableSparseMatrix(SparseMatrix):
+    def __mul__(self, other):
+        if isinstance(other, (InteractableDenseMatrix, InteractableSparseMatrix)):
+            return self.multiply(other)
+        elif isinstance(other, InteractableCategoricalMatrix):
+            return other.__mul__(self)
+        elif isinstance(other, (int, float)):
+            return self.multiply(numpy.array(other))
+        else:
+            raise TypeError(f"Cannot multiply {type(self)} and {type(other)}")
+        # Multiplication with categorical is handled by the categorical
+
+    def __rmul__(self, other):
+        return self.__mul__(other)
+
+    def to_non_interactable(self):
+        return SparseMatrix(self)
+
+    def get_names(self, col):
+        return [col]
+
+
+class InteractableCategoricalMatrix(CategoricalMatrix):
+    def __init__(self, *args, **kwargs):
+        multipliers = kwargs.pop("multipliers", None)
+        super().__init__(*args, **kwargs)
+        if multipliers is None:
+            self.multipliers = numpy.ones_like(self.cat, dtype=numpy.float_)
+        else:
+            self.multipliers = multipliers
+
+    def __mul__(self, other):
+        if isinstance(other, (InteractableDenseMatrix, float, int)):
+            result = copy.copy(self)
+            result.multipliers = result.multipliers * numpy.array(other)
+            return result
+        elif isinstance(other, InteractableSparseMatrix):
+            result = copy.copy(self)
+            result.multipliers = result.multipliers * other.todense()
+            return result
+        elif isinstance(other, InteractableCategoricalMatrix):
+            return self._interact_categorical(other)
+        else:
+            raise TypeError(
+                f"Can't multiply InteractableCategoricalMatrix with {type(other)}"
+            )
+
+    def __rmul__(self, other):
+        if isinstance(other, InteractableCategoricalMatrix):
+            other._interact_categorical(self)  # order matters
+        else:
+            return self.__mul__(other)
+
+    def to_non_interactable(self):
+        if numpy.all(self.multipliers == 1):
+            return CategoricalMatrix(
+                self.cat,
+                drop_first=self.drop_first,
+                dtype=self.dtype,
+            )
+        else:
+            return SparseMatrix(
```

Comment:
Not sure yet if a sparse matrix is the best way to go here if the multipliers come from a dense matrix. In that case, there should be a constant value for number of categories in `self` under which dense representations should be preferred, if I am not mistaken. But this can be addressed in a future iteration.

Reply:
Absolutely. There are also a couple other performance improvements that could be implemented in the future.